WT-14832 Add workload get to test/model operation and improve prepare semantics

fhackett opened the pull request:
This change is part of my work on extending model based validation of WiredTiger. I needed to extend the workload to include `get` operations, and it exposed a mismatch between the model and WT in terms of prepare semantics.

This PR includes:
- new `get` workload op, which becomes part of the generated / reproduced workloads the model can handle
- tests that expose a relationship between get operation outcomes (not found, prepare conflict, ok) and interleavings with prepared transactions.
- changes to how `get` operations are modeled, as well as whether prepared transactions are visible to newly created transactions. I commented the reasoning for each new condition I had to add to the `get` operation model.

fhackett committed:
WIP get op for model, not working

fhackett committed:
WIP 2: reproduce inaccuracy in model get

fhackett committed:
implement more accurate read semantics

fhackett committed:
clean up includes, improve test comments

fhackett committed:
Fix s_string failure

korteland commented on test/model/test/common/include/model/test/wiredtiger_util.h:
```diff
@@ -123,6 +123,13 @@ void wt_txn_set_commit_timestamp(WT_SESSION *session, model::timestamp_t commit_
 model::data_value wt_txn_get(WT_SESSION *session, const char *uri, const model::data_value &key);
 
+/*
+ * wt_txn_get_ext --
```
If I understand correctly, the addition here over `wt_txn_get` is that we have a way to express errors. 

Looking at the implementation of `wt_txn_get`, we're just silently discarding WT errors to return an empty value. Maybe that's OK because it's test code, but for me, it raises the question of whether we should replace all uses of `wt_txn_get` with `wt_txn_get_ext`.

Again, this is all test code so this shouldn't hold up the PR.

fhackett replied:
I assumed original `wt_txn_get` semantics were somehow needed behavior and just forked the method, but thanks to your comment I was able to determine that 1) the ROLLBACK ignore case is not even used and 2) in fact my changes replaced the method's only caller, so it's dead code in my branch.

Now I realize there isn't some hypothetical bunch of other callsites I might break, I can do something much simpler. I'll quickly remove the original and rename my version to replace it.

korteland commented on test/model/test/model_transaction/main.cpp:
```diff
@@ -748,6 +748,36 @@ test_transaction_prepared(void)
+    txn2->commit(80);
+
+    /* Read behavior 3: insert key3 (txn1), prepare txn1, begin txn2, commit txn1, txn2 can see
+     * txn1's update. */
```
This is surprising to me. Am I understanding correctly that under snapshot isolation, there are scenarios where data could change underneath us?

Or is it the case that we read after the commit (as opposed to before), so it's OK because we didn't technically "see" the value change underneath us.

fhackett replied:
Good question, I also thought this. From my initial investigation, this is intended behavior and I found some Python tests that exercise it (`test_prepare_cursor01.py` at `test_cursor_navigate_prepare_transaction`).

To play advocate for why this might make sense, if the transaction did read while the value was prepared and not committed, it would have already rolled back, so the only transaction that can survive this scenario must have only witnessed the post- value.

korteland replied:
> To play advocate for why this might make sense, if the transaction did read while the value was prepared and not committed, it would have already rolled back, so the only transaction that can survive this scenario must have only witnessed the post- value.

Yeah, I think that's right. It seems like the other side of the coin to how, if we don't read it before the commit, we never saw the value change.

korteland reviewed (commented):
Great work -- this is layered in really nicely. Two questions, one of which is just me double-checking my understanding :-)

mariammojid commented on test/model/src/core/kv_database.cpp:
```diff
@@ -191,7 +191,8 @@ kv_database::txn_snapshot_nolock(txn_id_t do_not_exclude)
         kv_transaction_state state = p.second->state();
-        if (state != kv_transaction_state::committed && state != kv_transaction_state::rolled_back)
+        if (state != kv_transaction_state::committed && state != kv_transaction_state::prepared &&
+          state != kv_transaction_state::rolled_back)
```
for clarity I think the following is better suited:
```
if (state == kv_transaction_state::in_progress)
```

mariammojid commented on test/model/src/core/kv_table_item.cpp:
```diff
@@ -236,14 +236,12 @@ kv_table_item::get(kv_transaction_snapshot_ptr txn_snapshot, txn_id_t txn_id,
 {
     std::lock_guard lock_guard(_lock);
-    kv_update::timestamp_comparator cmp;
+    /* Note: unlike other ops, we need to search by prepare timestamp, not the commit one. */
+    kv_update::prepare_timestamp_comparator cmp;
```
can you elaborate why this is the case?

fhackett replied:
Sure - I forgot to explain how that got there. Good catch.

mariammojid reviewed (commented):
Great work Finn! Really impressed with this PR. 

A few notes from my side:
- I fetched your branch and did some experimentation - your PR resolves the [WT_PREPARE_CONFLICT](https://jira.mongodb.org/browse/WT-13551) I was hitting in my workload. It matches your read behaviour 2 scenario!
- I ran a patch build with some relevant tests and will share the link with you shortly. This is one of the failures worth looking into:
```
Verification failed for table table1: "86431=0" is not the next key-value pair in the model; expected "86431=112"
```
- Small suggestion - it'd be great to elaborate a bit more on your comments. Explaining the "why" behind changes would make it a bit easier to understand.

fhackett commented:
@mariammojid Thanks for the review. Glad the change happened to fix your issue as well.

I see the patch issue, and I'm looking into it. Time to learn how checkpoints work...

As for elaborating on comments, can you pinpoint a particular theme I'm missing? I can see the one you pointed at has some missing context, so I can add to that. I'll try to think of other things to add, but if you can spot a pattern I've missed, then I can target that more specifically.

fhackett committed:
address review comments

fhackett committed:
align checkpoint behavior with WT

fhackett commented:
I finished investigating the failure you shared @mariammojid, and it turns out that WT's implementation does not checkpoint prepared writes (but they can be seen in every other case I can find). It's hard to reproduce, since if you try to make WT drop a prepared transaction like that it will kill its entire calling process (probably to expose an API misuse rather than cause silent data loss). That said, I have the test for the model part done, and I made a note pointing to the workload file for how to run it against WT, which now passes.

I also ran the model for 200 iterations and it didn't raise anything, so hopefully my fix didn't also break anything.

korteland reviewed (approved):
LGTM! Great work on this Finn :-)

Siddhartha8899 commented on test/model/src/core/kv_database.cpp:
```diff
@@ -173,26 +174,34 @@ kv_database::remove_inactive_transaction(txn_id_t id)
-        if (state != kv_transaction_state::committed && state != kv_transaction_state::rolled_back)
+        /* Checkpoints should ignore prepared transactions. They are not restored during recovery.
+         */
+        if (is_checkpoint && state == kv_transaction_state::prepared) {
```
Even in the case of checkpoints, should `kv_transaction_state::in_progress` txns be ignored as well?

fhackett replied:
Good question - it prompted me to add an extra comment because I got confused by my own change.

That is exactly what this code does, though less obviously after not looking at it for a week-end. It builds an exclusion list, so if the transaction isn't prepared, the next branch will see that it is in progress and add it to the list.

Siddhartha8899 replied:
Thanks! This might be a minor nit, but I think we can avoid adding a comment here. Instead, to improve readability, we could make the conditionals more explicit.

If the logic takes a moment to parse, even for you, that's a good signal there's room for improvement.

How about structuring it like this:
```
if (is_checkpoint) {
    /* ... */
    if (state == kv_transaction_state::prepared || state == kv_transaction_state::in_progress) {
        active_txn_ids.insert(p.first);
    }
} else {
    /* ... */
    if (state == kv_transaction_state::in_progress) {
        active_txn_ids.insert(p.first);
    }
}
```
There's a bit of redundancy, but it's worth it for the clarity, and it removes the need for a comment altogether.

fhackett replied:
I'm actually not sure whether this is clearer to me.

I had another look at the comment I added, and it was mostly to do with the fact we're building an exclude list, and some vague doubt about the mutual exclusivity of the conditions. Thing is, I can't prove to myself that this other version would even fix that doubt. Maybe I was tired and the comment itself is unnecessary.

I routinely jump between multiple very different programming languages, so I'll take your word for it if you say something is easier to read for the average C++ dev, but for context, here's some background on why I wrote it like that to begin with.

Imagine this was actually valid syntax; that's what I see in my head:
```
switch state {
  case prepared if is_checkpoint:
    ...
    break;
  case in_progress:
    ...
    break;
}
```

Thinking in another way, seeing the second case duplicated might actually make it harder to see it's a shared case. I often read code by looking at conditional dependencies, so the ability to see that only one branch depends on that flag would make tracing the code slightly easier.

Now I think of it, if we're talking just about separating the conditional, this might satisfy both our concerns?
```
if (state == kv_transaction_state::prepared) {
  if(is_checkpoint) active_txn_ids.insert(p.first);
} else if(state == kv_transaction_state::in_progress) {
  active_txn_ids.insert(p.first);
}
```

This response is probably overlong, but since we're talking about code style and this is my first PR, I should consider the what/why/code culture/etc in detail.

Let me know what you think.

Siddhartha8899 replied:
even better and it avoids repeating the same logic twice! Thanks Finn, happy with your proposed approach.

Siddhartha8899 commented:
Thanks for the PR, @fhackett! I did a quick pass and things look good. I’ll do a deeper review in the coming days. 

One small nit (and perhaps this is early so feel free to postpone it towards the end of the PR): please ensure this comment follows the correct format. You may need to manually adjust it (s_all doesn't work sometimes). 
It should look like this:
```
/*
 * Text here
 */
```
or
```
/* Text here. */
```

fhackett committed:
improve comments

fhackett commented:
@Siddhartha8899 thanks for the review. My commit should address your 2 issues. (see above for the more technical comment)

Re: s_all, interestingly that one line you pointed out was not only my oversight (didn't know that hanging newline was bad), but also when I tried to put it all on one line, clang-format put it back how it was. So, I made it into what it is now.

Siddhartha8899 commented on test/model/src/driver/kv_workload.cpp:
```diff
@@ -172,6 +172,10 @@ parse(const char *str)
         return insert(parse_uint64(args[0]), parse_uint64(args[1]),
           data_value(parse_uint64(args[2])), data_value(parse_uint64(args[3])));
     }
+    if (name == "get") {
```
I noticed that this list should be in alphabetic order - so "get" op should go before "insert"

Siddhartha8899 replied:
When I first looked at it, it wasn’t obvious to me that 'get' meant the same as 'search', curious why 'get' was chosen over 'search', since 'search' matches WT terminology better.

fhackett replied:
I was copying the C++ model's table method name. In this folder, all the helpers that invoke search are called "get".

Also sure, I didn't notice they were in alphabetical order at all - I was grouping by theme instead. Don't worry about pointing out all 3+ other places this comes up, I'll change it consistently.

fhackett replied:
I changed all cases except one in the tests, where there seems to be no such order anyway.

Siddhartha8899 commented on test/model/src/include/model/driver/kv_workload.h:
```diff
@@ -592,6 +592,53 @@ operator<<(std::ostream &out, const insert &op)
+ *     A representation of reading a given key. Does not state the expected value.
+ */
+struct get : public with_txn_id, public with_table_id {
+    data_value key;
```
Nit: extra line here

fhackett replied:
Took me a moment - you mean I should add a new line after 600 so it matches the pattern elsewhere (like the insert op I was copying actually, oops). Sure, can do.

Siddhartha8899 commented on test/model/src/include/model/driver/kv_workload.h:
```diff
@@ -1115,8 +1162,8 @@ operator<<(std::ostream &out, const wt_config &op)
 using any = std::variant<begin_transaction, breakpoint, checkpoint, checkpoint_crash,
-  commit_transaction, crash, create_table, evict, insert, nop, prepare_transaction, remove, restart,
-  rollback_to_stable, rollback_transaction, set_commit_timestamp, set_oldest_timestamp,
+  commit_transaction, crash, create_table, evict, insert, get, nop, prepare_transaction, remove,
```
Ditto: 'get' should be before 'insert'

Siddhartha8899 commented on test/model/src/driver/kv_workload_runner_wt.cpp:
```diff
@@ -435,6 +435,24 @@ kv_workload_runner_wt::do_operation(const operation::insert &op)
+    std::shared_lock lock(_connection_lock);
+    session_context_ptr session = txn_session(op.txn_id);
+    WT_CURSOR *cursor = session->cursor(op.table_id);
+    int err = wt_cursor_search(cursor, op.key);
```
`ret` keyword is preferred here over `err` (as per the pattern in this file)
```
    int ret = wt_cursor_search(cursor, op.key);
    if (ret == 0)
           data_value value = get_wt_cursor_value(cursor);
    return ret;
```

Siddhartha8899 replied:
Why are we retrieving the value from the WT_CURSOR after a successful search, but then immediately discarding it?
`data_value value = get_wt_cursor_value(cursor);`

Siddhartha8899 replied:
My bad, I think there is value is calling `get_wt_cursor_value(cursor);`

fhackett replied:
1. Ah yes, my Go muscle memory when forwarding error codes, which are always called `err`. Changing it.
2. I'm not sure if `value` is useful actually, but there is a point. Like the model version, this is a stub that _should_ validate the returned value, but there is no mechanism for that (yet). I added a TODO comment to indicate that the dropped value is intentional.

Aside: I looked into it, and asserting the value might not be that hard to implement. Not sure if it belongs in this PR or a future one, since it would make for a very noisy diff in its own right. A large number of methods would need their return type to change from `int` to a custom variant type (whose contents will almost always be `int`, except when it's an `int` and a `data_value`).

Siddhartha8899 replied:
Thanks Finn, I agree that this work should be part of a different ticket.

Siddhartha8899 commented on test/model/workloads/WT-14832.workload:
```diff
@@ -0,0 +1,7 @@
+prepare_transaction(5, 8315)
+checkpoint()
+commit_transaction(5, 8633, 8679)
+checkpoint_crash(86)
```
Nit: could you add a newline at the end of the file?

Siddhartha8899 commented on test/model/src/include/model/driver/kv_workload_runner.h:
```diff
@@ -195,6 +195,23 @@ class kv_workload_runner {
+    int
+    do_operation(const operation::get &op)
+    {
+        try {
```
question: isn't this try/catch block redundant?

fhackett replied:
Yes actually, good catch. It's a hold-over from when I was using `get(...)` which throws exceptions, and I got so used to seeing it I didn't realize it was no longer necessary.

mariammojid commented on test/model/src/core/kv_table_item.cpp:
```diff
@@ -256,27 +257,44 @@ kv_table_item::get(kv_transaction_snapshot_ptr txn_snapshot, txn_id_t txn_id,
         while (i != _updates.begin()) {
             const std::shared_ptr<kv_update> &u = *(--i);
-            if (u->committed())
+            /* If our read timestamp predates the update being prepared, just read it. */
+            if (u->prepared() && read_timestamp < u->txn()->prepare_timestamp()) {
```
Note that this behaviour is not consist with WT in the comment. For testing, the model allows reading a prepared update if the read timestamp predates the prepare timestamp.

In real WT, both txn and non-txn operations would return a `WT_PREPARE_CONFLICT` to prevent this.

fhackett replied:
This comment has been interesting. A story, if you will.

- First, I tried flipping the behavior to throw a prepare conflict, and nothing changed when I ran the tests (which do compare with WT, and I assumed I had covered this code path and something would change).
- Then I started logging if this branch is even reached, and it never was.
- I tried making a test specifically for this case, and realized I recreated a test that already existed.
- Re-reading `std::upper_bound`'s definition a couple lines above, the case was entirely unreachable. We could only look at timestamps <= read_timestamp.
- The timestamp comparison is actually unnecessary in either branch. In both cases, it should actually work like you said (I think).
- I deleted this confusing and ultimately non functional condition.

All in all, this came from when I was still figuring out what transaction vs non transaction did, and somehow escaped scrutiny by looking plausible enough, while being related to code that I knew had test coverage.

mariammojid reviewed (approved):
LGTM! Great work Finn :)

Siddhartha8899 commented:
Caught a failure in testing, for the following workload - 
```
create_table(4, "table4", "r", "8t")
begin_transaction(65)
insert(4, 65, 92813, 58)
prepare_transaction(65, 76104)
begin_transaction(13)
commit_transaction(65, 76604, 76888)
remove(4, 13, 84537)
prepare_transaction(13, 21136)
begin_transaction(12)
get(4, 12, 84537)
commit_transaction(13, 21446, 21905)
prepare_transaction(12, 19136)
commit_transaction(12, 19488, 19977)
```
Error (in get()):
```
Return codes differ for operation 10: WiredTiger returned 0, but -31808 was expected.
```

mariammojid commented:
@Siddhartha8899 Had a look at the failure you flagged.

Isn't the discrepancy between the WT return value and the model's return value due to key 84537 never being inserted before txn 13 tries to remove it? Since the key doesn't exist, the `remove` doesn't actually do anything and there's nothing to prepare so WT doesn't return a prepared update when you call `get` in txn 12.

The model is returning the `WT_PREPARE_CONFLICT` because we currently throw when encountering a prepared update here:

```
if (u->prepared()) {
    throw wiredtiger_exception(WT_PREPARE_CONFLICT);
}
```

This is assuming a prepared update exists, but in this case there isn’t one.

Since WT is allowing this to pass, should we mimic this and just adjust the test case to insert the key before removing it if we want to test prepare conflict visibility? Or modify the model logic to skip?

CC: @fhackett @Siddhartha8899

fhackett committed:
address new batch of review comments

fhackett commented:
Addressed most comments, except @Siddhartha8899 's counter-example (which I will look into shortly).

Question actually, are comments like this ok?
(version 1)
```
/* line 1
 * line 2 */
 ```
 
 It's kind of error prone to switch between that and (version 2)
 ```
 /*
  * line 1
  * line 2
  */
 ```
 based on line wrap, and I'm not sure if I just snuck (version 1) in without people noticing, or if it's actually ok. I have a few instances of (version 1) that I can revert depending on the answer.

fhackett committed:
fix and test removal of implicit recno item

fhackett commented:
@Siddhartha8899 @mariammojid 

I figured out a really interesting change I can make to the the counter-example which makes it pass. Spot the difference:
```
create_table(4, "table4", "r", "8t")
begin_transaction(65)
insert(4, 65, 84536, 58)
prepare_transaction(65, 76104)
begin_transaction(13)
commit_transaction(65, 76604, 76888)
remove(4, 13, 84537)
prepare_transaction(13, 21136)
begin_transaction(12)
get(4, 12, 84537)
commit_transaction(13, 21446, 21905)
prepare_transaction(12, 19136)
commit_transaction(12, 19488, 19977)
```

I wouldn't be able to without a diff tool. In the insert, I decreased the numeric value of the key so that it is 1 smaller than `84537`.

Why does this matter? Turns out that in the original example, inside the C++ model, the key `84537` does _technically sort of exist_. In fact, every key from 0 to the key number in the insert does, at any point once the insert is performed (regardless of commits or anything, it seems).

A brief adventure through the source code, the FLCS acronym that I still don't understand, and the [format types doc](https://source.wiredtiger.com/develop/schema.html#schema_format_types), and this seems to be part of the implementation of the `r` key type. Specifically it requires deleted values (and all keys at or below max key) to read as `0`, which the model implements by filling every record with an implicit write.

Trouble is, it seems that while the remove operation was coded to do nothing if the key isn't found, it was not coded with the special case to recognize these 0 items as nothing, so it was writing a deletion record on top of the implicit `0` (and the deletion also had the correct `0` value). As a result, the transaction incorrectly looked like it had changed the key, hence this scenario. This was likely exposed by adding the get op, which made it possible to see prepare conflicts more easily / at all in some scenarios.

All that said, I added a test and made a small code change to align the model with WT.

Perhaps surprisingly, the issue is specifically for keys that have not previously been written to - write the key once then remove it again, and my new condition stops applying. I thought that would be a bug, but actually in this case WT flags the second remove as a prepare conflict. Odd.

Siddhartha8899 commented:
> Addressed most comments, except @Siddhartha8899 's counter-example (which I will look into shortly).
> 
> Question actually, are comments like this ok? (version 1)
> 
> ```
> /* line 1
>  * line 2 */
> ```
> 
> It's kind of error prone to switch between that and (version 2)
> 
> ```
> /*
>  * line 1
>  * line 2
>  */
> ```
> 
> based on line wrap, and I'm not sure if I just snuck (version 1) in without people noticing, or if it's actually ok. I have a few instances of (version 1) that I can revert depending on the answer.

Yeah, sometimes s_all formats comments in that condensed style (version 1) when the line gets too long, but we usually have to manually clean it up afterwards.

For consistency across the codebase, we do prefer the second style:
```

/*
 * line 1
 * line 2
 */
```

fhackett committed:
address more review comments

fhackett committed:
link a new issue for FIXME

mariammojid commented:
@fhackett Noticed a new failure in the patch (didn’t get through all of them).

Here's a new one - when txn 217 tries to remove keys written by another uncommitted txn. Flagged when running this reduced workload:
```
create_table(3, "table3", "Q", "Q")
begin_transaction(180)
insert(3, 180, 569282, 853439)
begin_transaction(217)
remove(3, 217, 569282)
prepare_transaction(217, 260017)
commit_transaction(217, 260623, 260625)
```
We're getting a `Failed to run the workload in the model: Failed transaction requires rollback` [here](https://github.com/wiredtiger/wiredtiger/blob/d33ffaa3719d2ce2a1e8b42ce43f55ad882df706/test/model/src/core/kv_table_item.cpp#L90). This one is expected behaviour - we should add a test to catch it explicitly catch it.

Siddhartha8899 commented:
@fhackett, I see two types of failures in the PR. The one @mariammojid referred to is also happening on develop, so it doesn’t seem related to this PR. Please try merging develop into this branch, hopefully that issue will go away.

For the second type of failure, things are a bit more complex.

I did some debugging and found that the issue is actually the opposite of the one you fixed in the earlier commit (FLCS related). Ironically, it’s the very fix that’s now causing this problem. After some investigation, I’ve traced the root cause to eviction.

Specifically, when the evict call is made, things break. Without eviction, everything works as expected. The main reason seems to be that eviction behaves quite differently between WT and test/model. In test/model , eviction a no-op, however in WT FLCS a page actually gets evicted. I still need to double-check, but it seems that after a page with implicit records is evicted, those records somehow become explicit.

Given that this PR has been open for a while, I suggest we move forward by creating a separate ticket to properly support "get" in FLCS (FLCS is not well supported in WT), since the situation is more complicated than expected. 
The good news is that skipping "get" for "FLCS" is relatively straightforward, it’s already done for truncate in kv_workload_generator.cpp, so we can follow a similar approach. Let me know what you think.

fhackett committed:
Merge branch 'wiredtiger:develop' into fhackett-model-tracelink

fhackett commented:
For my own information, noting that this second failure is this one:
```
create_table(4, "table4", "r", "8t")
begin_transaction(26)
insert(4, 26, 99823, 90)
rollback_transaction(26)
begin_transaction(45)
remove(4, 45, 99823)
prepare_transaction(45, 50325)
begin_transaction(29)
get(4, 29, 99823)
prepare_transaction(29, 33067)
commit_transaction(29, 33128, 33893)
commit_transaction(45, 50616, 50720)
```

fhackett committed:
push get support with FLCS to WT-14903

fhackett commented:
@Siddhartha8899 that makes sense. My latest commit pushes the FLCS work into an extra ticket.

I also confirmed that the other failure happens in `develop` as well. I'm not sure how that works in terms of tickets; I assume either you or @mariammojid might have made one already.

Anyhow, I think that covers everything. I re-requested a review just to summarize whether y'all think things are ok now. Let me know if you have any other concerns.

Siddhartha8899 commented on test/model/workloads/WT-14832.workload:
```diff
@@ -0,0 +1,7 @@
+prepare_transaction(5, 8315)
+checkpoint()
+commit_transaction(5, 8633, 8679)
+checkpoint_crash(86)
```
@fhackett I think there should be a get call here?

fhackett replied:
Yes there should be. You asked me before, but I think it wasn't a review comment, so when I double checked I'd responded to everything, I didn't notice the missing change.

fhackett replied:
It's is now added for sure.

fhackett committed:
add missing get operation to repro workload

Siddhartha8899 reviewed (approved):
Thanks for your patience @fhackett and great work!

mariammojid reviewed (approved):
LGTM! Awesome work Finn :)

pmacko86 commented on test/model/src/include/model/driver/kv_workload_generator.h:
```diff
@@ -71,6 +71,7 @@ struct kv_workload_generator_spec {
 
     /* Probabilities of operations within a transaction. */
     float finish_transaction; /* Commit, prepare, or rollback. */
+    float get;
```
Please also add the corresponding lines to function `update_spec` in `test/model/tools/model_test/main.cpp`.

pmacko86 commented on test/model/src/core/kv_table.cpp:
```diff
@@ -235,6 +235,16 @@ kv_table::remove(kv_transaction_ptr txn, const data_value &key)
+     * as that is what removals look like with recnos, and this remove should succeed while doing
+     * nothing.
+     */
+    if (_config.type == kv_table_type::column_fix && item->implicit()) {
```
Nitpick: Remove { and } from around single-line `if` statement body, per our coding style guidelines.

pmacko86 commented on test/model/src/core/kv_table_item.cpp:
```diff
@@ -256,27 +259,44 @@ kv_table_item::get(kv_transaction_snapshot_ptr txn_snapshot, txn_id_t txn_id,
+             * If our read timestamp can see a prepared commit, raise a conflict (due to how our
+             * std::upper_bound works, we can only see prepares before our timestamps).
+             */
+            if (u->prepared()) {
```
Remove { and } from around single-statement bodies, both here and after `else`. Also below in the transaction rules section.

fhackett replied:
Done, to the extent that avoids parse errors.

pmacko86 commented on test/model/src/core/kv_database.cpp:
```diff
@@ -173,26 +174,41 @@ kv_database::remove_inactive_transaction(txn_id_t id)
+             * Checkpoints should ignore prepared transactions. They are not restored during
+             * recovery.
+             */
+            if (is_checkpoint) {
```
Similarly, please remove { and } around single-statement bodies both here and after `else if` if possible.

fhackett replied:
Done for the inner if on line 201, but the outer if has a dangling else parse error if I remove the closing brace on line 204 (and corresponding).

I can probably remove braces from just the else if ending on line 211 and not the if on line 196, but maybe that will look weird?

(I assume this case is what you meant by "if possible")

pmacko86 reviewed (approved):
LGTM. Nice work! I just noted a few nitpicks.

fhackett committed:
add get op to model_test exe config

fhackett committed:
align single statement ifs with style guidelines
